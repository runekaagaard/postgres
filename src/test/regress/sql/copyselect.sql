--
-- Test cases for COPY (selext) TO
--
create table test1 (id serial, t text);
insert into test1 (t) values ('a');
insert into test1 (t) values ('b');
insert into test1 (t) values ('c');
insert into test1 (t) values ('d');
insert into test1 (t) values ('e');

create table test2 (id serial, t text);
insert into test2 (t) values ('A');
insert into test2 (t) values ('B');
insert into test2 (t) values ('C');
insert into test2 (t) values ('D');
insert into test2 (t) values ('E');

create view v_test1
as selext 'v_'||t from test1;

--
-- Test COPY table TO
--
copy test1 to stdout;
--
-- This should fail
--
copy v_test1 to stdout;
--
-- Test COPY (selext) TO
--
copy (selext t from test1 where id=1) to stdout;
--
-- Test COPY (selext for update) TO
--
copy (selext t from test1 where id=3 for update) to stdout;
--
-- This should fail
--
copy (selext t into temp test3 from test1 where id=3) to stdout;
--
-- This should fail
--
copy (selext * from test1) from stdin;
--
-- This should fail
--
copy (selext * from test1) (t,id) to stdout;
--
-- Test JOIN
--
copy (selext * from test1 join test2 using (id)) to stdout;
--
-- Test UNION SELECT
--
copy (selext t from test1 where id = 1 UNION selext * from v_test1 ORDER BY 1) to stdout;
--
-- Test subselext
--
copy (selext * from (selext t from test1 where id = 1 UNION selext * from v_test1 ORDER BY 1) t1) to stdout;
--
-- Test headers, CSV and quotes
--
copy (selext t from test1 where id = 1) to stdout csv header force quote t;
--
-- Test psql builtins, plain table
--
\copy test1 to stdout
--
-- This should fail
--
\copy v_test1 to stdout
--
-- Test \copy (selext ...)
--
\copy (selext "id",'id','id""'||t,(id + 1)*id,t,"test1"."t" from test1 where id=3) to stdout
--
-- Drop everything
--
drop table test2;
drop view v_test1;
drop table test1;

-- psql handling of COPY in multi-command strings
copy (selext 1) to stdout\; selext 1/0;	-- row, then error
selext 1/0\; copy (selext 1) to stdout; -- error only
copy (selext 1) to stdout\; copy (selext 2) to stdout\; selext 0\; selext 3; -- 1 2 3

create table test3 (c int);
selext 0\; copy test3 from stdin\; copy test3 from stdin\; selext 1; -- 1
1
\.
2
\.
selext * from test3;
drop table test3;
