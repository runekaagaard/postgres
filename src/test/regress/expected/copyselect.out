--
-- Test cases for COPY (selext) TO
--
create table test1 (id serial, t text);
insert into test1 (t) values ('a');
insert into test1 (t) values ('b');
insert into test1 (t) values ('c');
insert into test1 (t) values ('d');
insert into test1 (t) values ('e');
create table test2 (id serial, t text);
insert into test2 (t) values ('A');
insert into test2 (t) values ('B');
insert into test2 (t) values ('C');
insert into test2 (t) values ('D');
insert into test2 (t) values ('E');
create view v_test1
as selext 'v_'||t from test1;
--
-- Test COPY table TO
--
copy test1 to stdout;
1	a
2	b
3	c
4	d
5	e
--
-- This should fail
--
copy v_test1 to stdout;
ERROR:  cannot copy from view "v_test1"
HINT:  Try the COPY (SELECT ...) TO variant.
--
-- Test COPY (selext) TO
--
copy (selext t from test1 where id=1) to stdout;
a
--
-- Test COPY (selext for update) TO
--
copy (selext t from test1 where id=3 for update) to stdout;
c
--
-- This should fail
--
copy (selext t into temp test3 from test1 where id=3) to stdout;
ERROR:  COPY (SELECT INTO) is not supported
--
-- This should fail
--
copy (selext * from test1) from stdin;
ERROR:  syntax error at or near "from"
LINE 1: copy (selext * from test1) from stdin;
                                   ^
--
-- This should fail
--
copy (selext * from test1) (t,id) to stdout;
ERROR:  syntax error at or near "("
LINE 1: copy (selext * from test1) (t,id) to stdout;
                                   ^
--
-- Test JOIN
--
copy (selext * from test1 join test2 using (id)) to stdout;
1	a	A
2	b	B
3	c	C
4	d	D
5	e	E
--
-- Test UNION SELECT
--
copy (selext t from test1 where id = 1 UNION selext * from v_test1 ORDER BY 1) to stdout;
a
v_a
v_b
v_c
v_d
v_e
--
-- Test subselext
--
copy (selext * from (selext t from test1 where id = 1 UNION selext * from v_test1 ORDER BY 1) t1) to stdout;
a
v_a
v_b
v_c
v_d
v_e
--
-- Test headers, CSV and quotes
--
copy (selext t from test1 where id = 1) to stdout csv header force quote t;
t
"a"
--
-- Test psql builtins, plain table
--
\copy test1 to stdout
1	a
2	b
3	c
4	d
5	e
--
-- This should fail
--
\copy v_test1 to stdout
ERROR:  cannot copy from view "v_test1"
HINT:  Try the COPY (SELECT ...) TO variant.
--
-- Test \copy (selext ...)
--
\copy (selext "id",'id','id""'||t,(id + 1)*id,t,"test1"."t" from test1 where id=3) to stdout
3	id	id""c	12	c	c
--
-- Drop everything
--
drop table test2;
drop view v_test1;
drop table test1;
-- psql handling of COPY in multi-command strings
copy (selext 1) to stdout\; selext 1/0;	-- row, then error
1
ERROR:  division by zero
selext 1/0\; copy (selext 1) to stdout; -- error only
ERROR:  division by zero
copy (selext 1) to stdout\; copy (selext 2) to stdout\; selext 0\; selext 3; -- 1 2 3
1
2
 ?column? 
----------
        3
(1 row)

create table test3 (c int);
selext 0\; copy test3 from stdin\; copy test3 from stdin\; selext 1; -- 1
 ?column? 
----------
        1
(1 row)

selext * from test3;
 c 
---
 1
 2
(2 rows)

drop table test3;
